stabilizer.py

```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/python
+"""
+Author: Vertigo Designs, Ryan Summers
+
+Description: Provides an API for controlling Stabilizer over Miniconf (MQTT).
+"""
+import argparse
+import asyncio
+import json
+
+from gmqtt  import Client as MqttClient
+
+
+def parse_value(value):
+    """ Parse a command-line value into the most appropriate associated python datatype. """
+    # If the value is an array, construct it as such and recurse for individual elements.
+    if value.startswith('[') and value.endswith(']'):
+        all_values = []
+        for data in value[1:][:-1].split(','):
+            all_values.append(parse_value(data))
+        return all_values
+
+    if value.isnumeric():
+        return int(value)
+    try:
+        return float(value)
+    except ValueError:
+        return value
+
+
+class MiniconfApi:
+    """ An asynchronous API for controlling Miniconf devices using the MQTT control interface. """
+
+    @classmethod
+    async def create(cls, identifier, broker):
+        """ Create a connection to MQTT for communication with the device. """
+        client = MqttClient(client_id='')
+        await client.connect(broker)
+        return cls(client, identifier)
+
+
+    def __init__(self, client, identifier):
+        """ Consructor.
+
+        Args:
+            client: A connected MQTT5 client.
+            identifier: The ID of the device to control.
+        """
+        self.response_topic = f'{identifier}/feedback'
+        self.client = client
+        self.identifier = identifier
+        self.command_complete = asyncio.Event()
+        self.client.on_message = self._handle_response
+        self.response = None
+
+        self.client.subscribe(self.response_topic)
+
+
+    def _handle_response(self, client, topic, payload, *_args, **_kwargs):
+        """ Callback function for when messages are received over MQTT.
+
+        Args:
+            client: The MQTT client.
+            topic: The topic that the message was received on.
+            payload: The payload of the message.
+        """
+        if topic != self.response_topic:
+            raise Exception(f'Unknown topic: {topic}')
+
+        # Indicate a response was received.
+        self.response = payload.decode('ascii')
+        self.command_complete.set()
+
+
+    async def _command(self, topic, message):
+        """ Send a command to a topic.
+
+        Args:
+            topic: The topic to send the message to.
+            message: The message to send to the provided topic.
+
+        Returns:
+            The received response to the command.
+        """
+        self.command_complete.clear()
+        self.client.publish(topic, payload=message, qos=0, retain=False,
+                            response_topic=self.response_topic)
+        await self.command_complete.wait()
+
+        response = self.response
+        self.response = None
+
+        return response
```

Comment:
These should be a map from topic to `Future`s that you `set_result()` in `_handle_response`.
There can be multiple in flight with `asyncio`.

Reply:
Current design does it all sequentially, since the CLI only supports a single setting configuration at once - the key=value pairs are parsed into a single JSON object that is sent to a single settings message, unless I'm misunderstanding what you're asking?

Reply:
Yes. No problem if the only user is this CLI.
But other users of the API could fire multiple requests before the responses come in.

Reply:
In the interest of simplicity, I would say let's leave it for now, and potentially expand if we have that use case later on

Reply:
Ok. But then:

* Use a `Future` instead of the `Event`+`self.response`.
* Add documentation stating that the user needs to ensure only one request in flight.

Reply:
There's actually no data the the API can use to differentiate one response from another, so it's not possible to have a mapping. If we wanted to support multiple in-flight messages, we would have to do one of:
* Utilize fire-and-forget configurations
* Modify `miniconf` to add a `Property` to the response object indicating the setting that the response is associated with.

Reply:
You could have stabilizer reply on unique response topics or return a token. But restricting this to a single call in flight is OK for now. Just make sure it's documented.
That's one thing. Using `Future` instead of `Event` + `self.response` is the other comment, to make this use more typical `asyncio` patterns.